MonoGame.Framework/Graphics/Texture2D.cs

```diff
@@ -308,7 +294,56 @@ private static void ConvertToABGR(int pixelHeight, int pixelWidth, int[] pixels)
                 uint pixel = (uint)pixels[i];
                 pixels[i] = (int)((pixel & 0xFF00FF00) | ((pixel & 0x00FF0000) >> 16) | ((pixel & 0x000000FF) << 16));
             }
+        }
+
+        private void ValidateParams<T>(int level, int arraySlice, Rectangle? rect, T[] data,
+            int startIndex, int elementCount, out Rectangle checkedRect) where T : struct
+        {
+            var textureBounds = new Rectangle(0, 0, Math.Max(width >> level, 1), Math.Max(height >> level, 1));
+            checkedRect = rect.HasValue ? rect.Value : textureBounds;
+            if (level < 0 || level >= LevelCount)
+                throw new ArgumentException("level must be smaller than the number of levels in this texture.");
+            if (arraySlice > 0 && !GraphicsDevice.GraphicsCapabilities.SupportsTextureArrays)
+                throw new ArgumentException("Texture arrays are not supported on this graphics device", "arraySlice");
+            if (arraySlice < 0 || arraySlice >= ArraySize)
+                throw new ArgumentException("arraySlice must be smaller than the ArraySize of this texture and larger than 0.");
+            if (!textureBounds.Contains(checkedRect))
+                throw new ArgumentException("Rectangle must be inside the texture bounds", "rect");
+            if (data == null)
+                throw new ArgumentNullException("data");
+            var tSize = Marshal.SizeOf(typeof(T));
+            var fSize = Format.GetSize();
+            if (tSize > fSize || fSize % tSize != 0)
+                throw new ArgumentException("Type T is of an invalid size for the format of this texture.", "T");
+            if (startIndex < 0 || startIndex >= data.Length)
+                throw new ArgumentException("startIndex must be at least zero and smaller than data.Length.", "startIndex");
+            if (data.Length < startIndex + elementCount)
+                throw new ArgumentException("The data array is too small.");
+
+            int dataByteSize;
+            if (Format.IsCompressedFormat())
+            {
+                // round x and y down to next multiple of four; width and height up to next multiple of four
+                checkedRect = new Rectangle(checkedRect.X & ~0x3, checkedRect.Y & ~0x3,
+#if OPENGL
+                    // OpenGL only: The last two mip levels require the width and height to be
+                    // passed as 2x2 and 1x1, but there needs to be enough data passed to occupy
+                    // a 4x4 block.
+                    checkedRect.Width > 4 ? (checkedRect.Width + 3) & ~0x3 : checkedRect.Width,
+                    checkedRect.Height > 4 ? (checkedRect.Height + 3) & ~0x3 : checkedRect.Height);
+#else
+                    (checkedRect.Width + 3) & ~0x3, (checkedRect.Height + 3) & ~0x3);
+#endif
+                dataByteSize = checkedRect.Width * checkedRect.Height * fSize / 16;

```

Comment:
I should change something here... Since below we check `elementCount * tSize != dataByteSize`, but for OpenGL/DX the checkedRect will be a different size for the lower mips so they require a different elementCount.